File: Menu_Driven.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define RANGE 255
#include <math.h>
#include <limits.h> 
#include <stdbool.h> 
	int main()
{		int z;
        void insertionSort(int arr[], int n)
					{ 
    						int i, key, j; 
   						for (i = 1; i < n; i++) { 
        					key = arr[i]; 
        					j = i - 1; 
  
        
        					while (j >= 0 && arr[j] > key) { 
            						arr[j + 1] = arr[j]; 
            						j = j - 1; 
        					} 
        					arr[j + 1] = key; 
    						} 
					} 
  
 
				void printArray(int arr[], int n) 
					{ 
    						int i; 
    						for (i = 0; i < n; i++) 
        						printf("%d ", arr[i]); 
    							printf("\n"); 
					} 
		int search(int arr[], int n, int x) 
			{ 
    				int i; 
    				for (i = 0; i < n; i++) 
        				if (arr[i] == x) 
            					return i; 
    				return -1; 
			} 
		
		int binarySearch(int arr[], int l, int r, int x) 
			{ 
    				if (r >= l) { 
        				int mid = l + (r - l) / 2; 
   
        				if (arr[mid] == x) 
            					return mid; 
  
         
        				if (arr[mid] > x) 
            					return binarySearch(arr, l, mid - 1, x); 
  
     
        				return binarySearch(arr, mid + 1, r, x); 
    						} 
   
    			return -1; 
			}
		
		void merge(int arr[], int l, int m, int r)
			{
    				int i, j, k;
    				int n1 = m - l + 1;
    				int n2 = r - m;
 
    
    				int L[n1], R[n2];
 
        			for (i = 0; i < n1; i++)
        				L[i] = arr[l + i];
    				for (j = 0; j < n2; j++)
        				R[j] = arr[m + 1 + j];
 
    
    				i = 0; 
    				j = 0; 
    				k = l; 
    				while (i < n1 && j < n2) {
        				if (L[i] <= R[j]) {
            					arr[k] = L[i];
            					i++;
        				}
        				else {
            					arr[k] = R[j];
            					j++;
        				}
        				k++;
    				}
 
    
    
    				while (i < n1) {
        				arr[k] = L[i];
        				i++;
        				k++;
    				}
 
    
        			while (j < n2) {
        				arr[k] = R[j];
        				j++;
        				k++;
    				}
			}
 

			void mergeSort(int arr[], int l, int r)
				{
    					if (l < r) {
                				int m = l + (r - l) / 2;
 
        
        					mergeSort(arr, l, m);
        					mergeSort(arr, m + 1, r);
 
        					merge(arr, l, m, r);
    					}
				}
 

			void printarray(int A[], int size)
				{
    					int i;
    					for (i = 0; i < size; i++)
        					printf("%d ", A[i]);
    						printf("\n");
				}
				
			void swap(int* a, int* b) 
				{ 
    					int t = *a; 
    					*a = *b; 
    					*b = t; 
				} 
  

			int partition (int arr[], int low, int high) 
				{ 
    					int pivot = arr[high];    
    					int i = (low - 1);   
  
    					for (int j = low; j <= high- 1; j++) 
    					{ 
         
        					if (arr[j] < pivot) 
        						{ 
            							i++;    
            							swap(&arr[i], &arr[j]); 
        						} 
    					} 
    					swap(&arr[i + 1], &arr[high]); 
    					return (i + 1); 
				} 
  
			void quickSort(int arr[], int low, int high) 
				{ 
    					if (low < high) 
    					{ 
                				int pi = partition(arr, low, high); 
  
         
        					quickSort(arr, low, pi - 1); 
        					quickSort(arr, pi + 1, high); 
    					} 
				} 
  

			void PrintArray(int arr[], int size) 
				{ 
    					int i; 
    					for (i=0; i < size; i++) 
        					printf("%d ", arr[i]); 
    						printf("\n"); 
				} 
				
			  
            void counting_sort(int A[], int k, int n)  
            {  
                int i, j;  
                int B[15], C[100];  
                for (i = 0; i <= k; i++)  
                    C[i] = 0;  
                for (j = 1; j <= n; j++)  
                    C[A[j]] = C[A[j]] + 1;  
                for (i = 1; i <= k; i++)  
                    C[i] = C[i] + C[i-1];  
                for (j = n; j >= 1; j--)  
                {  
                    B[C[A[j]]] = A[j];  
                    C[A[j]] = C[A[j]] - 1;  
                }  
                printf("The Sorted array is : ");  
                for (i = 1; i <= n; i++)  
                    printf("%d ", B[i]);  
            }  
            /*  End of counting_sort()  */  
               
            /*  The main() begins  */  
  
	       
			void Search(char* pat, char* txt) 
{ 
    int M = strlen(pat); 
    int N = strlen(txt); 
  
    /* A loop to slide pat[] one by one */
    for (int i = 0; i <= N - M; i++) { 
        int j; 
  
        /* For current index i, check for pattern match */
        for (j = 0; j < M; j++) 
            if (txt[i + j] != pat[j]) 
                break; 
  
        if (j == M) // if pat[0...M-1] = txt[i, i+1, ...i+M-1] 
            printf("Pattern found at index %d \n", i); 
    } 
} 	
		
  	        void prefixSuffixArray(char* pat, int M, int* pps) {
   int length = 0;
   pps[0] = 0;
   int i = 1;
   while (i < M) {
      if (pat[i] == pat[length]) {
         length++;
         pps[i] = length;
         i++;
      } else {
         if (length != 0)
         length = pps[length - 1];
         else {
            pps[i] = 0;
            i++;
         }
      }
   }
}
                void KMPAlgorithm(char* text, char* pattern) {
   int M = strlen(pattern);
   int N = strlen(text);
   int pps[M];
   prefixSuffixArray(pattern, M, pps);
   int i = 0;
   int j = 0;
   while (i < N) {
      if (pattern[j] == text[i]) {
         j++;
         i++;
      }
      if (j == M) {
         printf("Found pattern at index %d\n", i - j);
         j = pps[j - 1];
      }
      else if (i < N && pattern[j] != text[i]) {
         if (j != 0)
         j = pps[j - 1];
         else
         i = i + 1;
      }
   }
}
  	        #define d 10

            void rabinKarp(char pattern[], char text[], int q) {
  int m = strlen(pattern);
  int n = strlen(text);
  int i, j;
  int p = 0;
  int t = 0;
  int h = 1;

  for (i = 0; i < m - 1; i++)
    h = (h * d) % q;

  // Calculate hash value for pattern and text
  for (i = 0; i < m; i++) {
    p = (d * p + pattern[i]) % q;
    t = (d * t + text[i]) % q;
  }

  // Find the match
  for (i = 0; i <= n - m; i++) {
    if (p == t) {
      for (j = 0; j < m; j++) {
        if (text[i + j] != pattern[j])
          break;
      }

      if (j == m)
        printf("Pattern is found at position:  %d \n", i + 1);
    }

    if (i < n - m) {
      t = (d * (t - text[i] * h) + text[i + m]) % q;

      if (t < 0)
        t = (t + q);
    }
  }
}
  	        #define V 5 
  
// A utility function to find the vertex with 
// minimum key value, from the set of vertices 
// not yet included in MST 
            int minKey(int key[], bool mstSet[]) 
{ 
    // Initialize min value 
    int min = INT_MAX, min_index; 
  
    for (int v = 0; v < V; v++) 
        if (mstSet[v] == false && key[v] < min) 
            min = key[v], min_index = v; 
  
    return min_index; 
} 
  
// A utility function to print the 
// constructed MST stored in parent[] 
            int printMST(int parent[], int graph[V][V]) 
{ 
    printf("Edge \tWeight\n"); 
    for (int i = 1; i < V; i++) 
        printf("%d - %d \t%d \n", parent[i], i, graph[i][parent[i]]); 
} 
  
// Function to construct and print MST for 
// a graph represented using adjacency 
// matrix representation 
            void primMST(int graph[V][V]) 
{ 
    // Array to store constructed MST 
    int parent[V]; 
    // Key values used to pick minimum weight edge in cut 
    int key[V]; 
    // To represent set of vertices included in MST 
    bool mstSet[V]; 
  
    // Initialize all keys as INFINITE 
    for (int i = 0; i < V; i++) 
        key[i] = INT_MAX, mstSet[i] = false; 
  
    // Always include first 1st vertex in MST. 
    // Make key 0 so that this vertex is picked as first vertex. 
    key[0] = 0; 
    parent[0] = -1; // First node is always root of MST 
  
    // The MST will have V vertices 
    for (int count = 0; count < V - 1; count++) { 
        // Pick the minimum key vertex from the 
        // set of vertices not yet included in MST 
        int u = minKey(key, mstSet); 
  
        // Add the picked vertex to the MST Set 
        mstSet[u] = true; 
  
        // Update key value and parent index of 
        // the adjacent vertices of the picked vertex. 
        // Consider only those vertices which are not 
        // yet included in MST 
        for (int v = 0; v < V; v++) 
  
            // graph[u][v] is non zero only for adjacent vertices of m 
            // mstSet[v] is false for vertices not yet included in MST 
            // Update the key only if graph[u][v] is smaller than key[v] 
            if (graph[u][v] && mstSet[v] == false && graph[u][v] < key[v]) 
                parent[v] = u, key[v] = graph[u][v]; 
    } 
  
    // print the constructed MST 
    printMST(parent, graph); 
} 
 	        int temp;  
  
            void heapify(int arr[], int size, int i)  
{  
int largest = i;    
int left = 2*i + 1;    
int right = 2*i + 2;    
  
if (left < size && arr[left] >arr[largest])  
largest = left;  
  
if (right < size && arr[right] > arr[largest])  
largest = right;  
  
if (largest != i)  
{  
temp = arr[i];  
    arr[i]= arr[largest];   
    arr[largest] = temp;  
heapify(arr, size, largest);  
}  
}  
  
            void heapSort(int arr[], int size)  
{  
int i;  
for (i = size / 2 - 1; i >= 0; i--)  
heapify(arr, size, i);  
for (i=size-1; i>=0; i--)  
{  
temp = arr[0];  
    arr[0]= arr[i];   
    arr[i] = temp;  
heapify(arr, i, 0);  
}  
}  
  
		printf("Choose the operation to be performed\n");
		printf("1. Insertion Sort\n");
		printf("2. Linear Search\n");
		printf("3. Binary Search\n");
		printf("4. Merge Sort\n");
		printf("5. Quick Sort\n");
		printf("6. Counting Sort\n");
		printf("7. Naive String \n");
		printf("8. KMP\n");
		printf("9. Rabin Karp\n");
		printf("10. MST (prims algorithm)\n");
		printf("11. Heap Sort\n");
		scanf("%d", &z);
		switch (z){
			case 1:
			     
					{ 
    						int arr[] = { 12, 11, 13, 5, 6 }; 
    						int n = sizeof(arr) / sizeof(arr[0]); 
  
    						insertionSort(arr, n); 
    						printArray(arr, n); 
  
    						return 0; 
					} 
				

				
			
			break;
		
		case 2:
			
  
	 
			
			{ 
    				int arr[] = { 2, 3, 4, 10, 40 }; 
    				int x = 10; 
    				int n = sizeof(arr) / sizeof(arr[0]); 
    
     
    				int result = search(arr, n, x); 
    				(result == -1) 
        				? printf("Element is not present in array") 
        				: printf("Element is present at index %d", result); 
    				return 0; 
			}
		case 3:
			
  
			
			{ 
    				int arr[] = { 2, 3, 4, 10, 40 }; 
    				int n = sizeof(arr) / sizeof(arr[0]); 
    				int x = 10; 
    				int result = binarySearch(arr, 0, n - 1, x); 
    				(result == -1) ? printf("Element is not present in array") 
                   		: printf("Element is present at index %d", 
                            		result); 
    				return 0; 
			} 

		case 4:
			

			
				{
    					int arr[] = { 12, 11, 13, 5, 6, 7 };
    					int arr_size = sizeof(arr) / sizeof(arr[0]);
 
    					printf("Given array is \n");
    					printarray(arr, arr_size);
 
    					mergeSort(arr, 0, arr_size - 1);
 
    					printf("\nSorted array is \n");
    					printarray(arr, arr_size);
    					return 0;
				}

		case 5:
			
 
				{ 
    					int arr[] = {10, 7, 8, 9, 1, 5}; 
    					int n = sizeof(arr)/sizeof(arr[0]); 
    					quickSort(arr, 0, n-1); 
    					printf("Sorted array: \n"); 
    					PrintArray(arr, n); 
    					return 0; 
				} 

		case 6:
				{  
                        int n, k = 0, A[15], i;  
                        printf("Enter the number of input : ");  
                        scanf("%d", &n);  
                        printf("\nEnter the elements to be sorted :\n");  
                        for (i = 1; i <= n; i++)  
                        {  
                            scanf("%d", &A[i]);  
                            if (A[i] > k) {  
                                k = A[i];  
                            }  
                        }  
                        counting_sort(A, k, n);  
                        printf("\n");  
                        return 0;  
}
		
		case 7:
		       {
		         char txt[] = "AABAACAADAABAAABAA"; 
                char pat[] = "AABA"; 
                Search( pat, txt); 
                return 0;         
 }
		case 8:
		{
		    char text[] = "xyztrwqxyzfg";
   char pattern[] = "xyz";
   printf("The pattern is found in the text at the following index : \n");
   KMPAlgorithm(text, pattern);
   return 0;
		}
		
		case 9:
		{
		     char text[] = "ABCCDDAEFG";
  char pattern[] = "CDD";
  int q = 13;
  rabinKarp(pattern, text, q);
		}
		
		case 10:
		{
		     int graph[V][V] = { { 0, 2, 0, 6, 0 }, 
                        { 2, 0, 3, 8, 5 }, 
                        { 0, 3, 0, 0, 7 }, 
                        { 6, 8, 0, 0, 9 }, 
                        { 0, 5, 7, 9, 0 } }; 
  
    // Print the solution 
    primMST(graph); 
  
    return 0; 
		}
		
		case 11:
		{
		    int arr[] = {1, 10, 2, 3, 4, 1, 2, 100,23, 2};  
int i;  
int size = sizeof(arr)/sizeof(arr[0]);  
  
heapSort(arr, size);  
  
printf("printing sorted elements\n");  
for (i=0; i<size; ++i)  
printf("%d\n",arr[i]);  
		}

		default:
			printf("Out of Range");
	}
	return 0; 
}

File: Menu_Driven.py
import sys
def choice():
    print("Select a Lab no.")
    print("1.Linear Search")
    print("2.Binary Search")
    print("3.Insertion Sort")
    print("4.erge Sort")
    print("5.Quick Sort")
    print("6.Counting Sort")
    print("7.Longest Common Subsequence")
    print("8.Naive String Matching")
    print("9.KMP")
    print("10.Minimum Spanning tree")
    print("11.Rabin-karp Matching")
    print("12.Heap Sort")
def insertion_sort(a):
    for j in range(2, len(a)):
        key = a[j]
        i = j - 1
        while (i >= 0 and key < a[i]):
            a[i + 1] = a[i]
            i = i - 1
        a[i + 1] = key
def linear_Search(list,n,key):  
    for i in range(0, n):  
        if (list[i] == key):  
            return i  
    return -1   
def binarySearch (arr, first, last, x): 
    if last>=first: 
        mid=int((first+last)/2)
        if arr[mid]==x: 
            return mid 
        elif arr[mid] > x: 
            return binarySearch(arr, first, mid-1, x) 
        else: 
            return binarySearch(arr, mid + 1, last, x) 
    else: 
        return -1
def merge(arr, l, m, r): 
    n1 = m - l + 1
    n2 = r- m  
    L = [0] * (n1) 
    R = [0] * (n2)  
    for i in range(0 , n1): 
        L[i] = arr[l + i] 
    for j in range(0 , n2): 
        R[j] = arr[m + 1 + j] 
    i = 0    
    j = 0      
    k = l 
    while i < n1 and j < n2 : 
        if L[i] <= R[j]: 
            arr[k] = L[i] 
            i += 1
        else: 
            arr[k] = R[j] 
            j += 1
        k += 1
    while i < n1: 
        arr[k] = L[i] 
        i += 1
        k += 1
    while j < n2: 
        arr[k] = R[j] 
        j += 1
        k += 1
def mergeSort(arr,l,r): 
    if l < r: 
        m = (l+(r-1))//2
        mergeSort(arr, l, m) 
        mergeSort(arr, m+1, r) 
        merge(arr, l, m, r)    
def partition(arr, low, high):
    i = (low-1)        
    pivot = arr[high]     
    for j in range(low, high):
        if arr[j] <= pivot:
            i = i+1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return (i+1)
def quickSort(arr, low, high):
    if len(arr) == 1:
        return arr
    if low < high:
        pi = partition(arr, low, high)
        quickSort(arr, low, pi-1)
        quickSort(arr, pi+1, high)  
def counting_sort(array1, max_val):
    m = max_val + 1
    count = [0] * m                 
    for a in array1:
        count[a] += 1             
    i = 0
    for a in range(m):            
        for c in range(count[a]):  
            array1[i] = a
            i += 1
    return array1  
def lcs(X , Y): 
    m = len(X) 
    n = len(Y) 
    L = [[None]*(n+1) for i in range(m+1)] 
    for i in range(m+1): 
        for j in range(n+1): 
            if i == 0 or j == 0 : 
                L[i][j] = 0
            elif X[i-1] == Y[j-1]: 
                L[i][j] = L[i-1][j-1]+1
            else: 
                L[i][j] = max(L[i-1][j] , L[i][j-1]) 
    return L[m][n] 
def naive(pat,txt):
    M = len(pat) 
    N = len(txt)
    for i in range(N - M + 1):
        s=0
        while(s < M): 
            if (txt[i + s] != pat[s]): 
                break
            s += 1
        if (s == M):  
            print("Pattern found with shift ", i) 
def array(P):
    m=len(P)
    k=-1
    pie=[int(i) for i in range(m)]
    pie[0]=-1
    for q in range (1,m):
        while (k>=0 and P[k+1]!=P[q]):
            k=pie[k]
        if P[k+1]==P[q]:
            k=k+1
        pie[q]=k
    for q in range (0,m):
        print(pie[q]+1,end=' ')
    
    return pie
def KMP(T,P):
    q=-1
    n=len(T)
    m=len(P)
    pie = array(P)
    for i in range(0,n):
        while (q>=0 and P[q+1]!=T[i]):
            q=pie[q]
        if P[q+1]==T[i]:
            q=q+1
        if q==m-1:
            print("Pattern occur at shift: ",i-m+1)
            q=pie[q]
class Graph(): 
  
    def _init_(self, vertices): 
        self.V = vertices 
        self.graph = [[0 for column in range(vertices)]  
                    for row in range(vertices)] 
    def printMST(self, parent): 
        print ("Edge \tWeight")
        for i in range(1, self.V): 
            print(parent[i], "-", i, "\t", self.graph[i][ parent[i]] )
    def minKey(self, key, mstSet): 
        min = sys.maxint 
  
        for v in range(self.V): 
            if key[v] < min and mstSet[v] == False: 
                min = key[v] 
                min_index = v 
  
        return min_index
    def primMST(self): 
        key = [sys.maxint] * self.V 
        parent = [None] * self.V 
        key[0] = 0 
        mstSet = [False] * self.V 
        for cout in range(self.V):  
            u = self.minKey(key, mstSet)
            mstSet[u] = True 
            for v in range(self.V): 
                if self.graph[u][v] > 0 and mstSet[v] == False and key[v] > self.graph[u][v]: 
                        key[v] = self.graph[u][v] 
                        parent[v] = u 
  
        self.printMST(parent)        
def search(pat, txt, q): 
    M = len(pat) 
    N = len(txt) 
    i = 0
    j = 0
    p = 0     
    t = 0     
    h = 1 
    for i in range(M-1): 
        h = (h*d)%q 
    for i in range(M): 
        p = (d*p + ord(pat[i]))%q 
        t = (d*t + ord(txt[i]))%q 
    for i in range(N-M+1): 
        if p==t:  
            for j in range(M): 
                if txt[i+j] != pat[j]: 
                    break
                else: j+=1 
            if j==M: 
                print ("\nPattern occur at shift " + str(i))
        if i < N-M: 
            t = (d*(t-ord(txt[i])*h) + ord(txt[i+M]))%q  
            if t < 0: 
                t = t+q      
            
def heapify(arr, n, i):
    largest = i  
    l = 2 * i + 1    
    r = 2 * i + 2     
    if l < n and arr[largest] < arr[l]:
        largest = l
    if r < n and arr[largest] < arr[r]:
        largest = r
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  
        heapify(arr, n, largest)
def heapSort(arr):
    n = len(arr)
    for i in range(n//2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)
      
def main(): 
    while True:
        choice()
        option =eval(input('Enter Choice from 1-14 from the menu or enter  to exit.'))
        if(option==3):
            print('Insertion Sort')
            a= input('Enter the list of numbers: ').split()
            a = [int(x) for x in a]
            insertion_sort(a)
            print('Sorted list: ',end=' ')
            print(a)
        elif(option==1):
            print('Linear Search')
            list= input('Enter the list of numbers: ').split()    
            list = [int(x) for x in list]
            key = int(input('Enter key: '))
            n = len(list)  
            res = linear_Search(list, n, key)  
            if(res == -1):  
                print("Element not found")  
            else:  
                print("Element found at index: ", res)
        elif(option==2):
            print('Binary Search')
            arr = input('Enter the list of numbers: ').split()
            arr=[int(i)for i in arr]
            x=int(input("Enter a key: "))
            result = binarySearch(arr, 0, len(arr)-1, x) 
            if result != -1: 
                print ("Element is present at index: % d" % result) 
            else: 
                print ("Element is not present in array") 
        elif(option==4):
            print("Merge Sort")
            arr=input("Enter the list of numbers: ").split()
            arr=[int(x) for x in arr] 
            mergeSort(arr,0,len(arr)-1) 
            print("Sorted list",end=' ')
            print(arr)
        elif(option==5):
            print("Quick Sort")
            arr = input("Enter the list of numbers: ").split()
            arr=[int(x) for x in arr] 
            n = len(arr)
            quickSort(arr, 0, n-1)
            print("Sorted array is:")
            for i in range(n):
                print("%d" % arr[i])
        elif(option==6):
            print("Counting Sort")
            arr = input("Enter the list of numbers: ").split()
            arr=[int(x) for x in arr] 
            m=max(arr)
            print(counting_sort(arr,m))
        elif(option==7):
            print('Longest Common Subsequence')
            X =input("Enter first string: ")
            print(X)
            Y =input('Enter second string: ')
            print(Y)  
            print ("Length of LCS is ", lcs(X, Y) )
        elif(option==8):
            print('Naive String matching')
            pat =input("Enter pattern string: ")
            txt =input('Enter text string: ')  
            print(naive(pat,txt))
        elif(option==9):
            T=input("Enter text: ")
            P=input("Enter pattern: ")
            n=len(T)
            m=len(P)
            KMP(T,P) 
        elif(option==10):
            g = Graph(5) 
            g.graph = [ [0, 2, 0, 6, 0], 
                        [2, 0, 3, 8, 5], 
                        [0, 3, 0, 0, 7], 
                        [6, 8, 0, 0, 9], 
                        [0, 5, 7, 9, 0]] 
  
            g.primMST(); 
        elif(option==11):
            txt = input("Enter Text: ")
            pat = input("Enter Pattern")
            d=10
            q = 13 
            search(pat,txt,q)
        elif(option==12):
            arr = input("Enter the list of numbers: ").split()
            arr=[int(x) for x in arr] 
            heapSort(arr)
            n = len(arr)
            print("Sorted array is")
            for i in range(n):
                print("%d" % arr[i],end=' ')
        elif option==13:
                break
        else:
            print("Invalid Lab")
main()